openshift-tools/pkg/discover-etcd-initial-cluster/initial-cluster_test.go

```diff
@@ -0,0 +1,143 @@
+package discover_etcd_initial_cluster
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/url"
+	"os"
+	"reflect"
+	"sort"
+	"strings"
+	"testing"
+
+	"go.etcd.io/etcd/integration"
+	"go.etcd.io/etcd/pkg/transport"
+)
+
+var testTLSInfo = transport.TLSInfo{
+	KeyFile:        "../../../integration/fixtures/server.key.insecure",
+	CertFile:       "../../../integration/fixtures/server.crt",
+	TrustedCAFile:  "../../../integration/fixtures/ca.crt",
+	ClientCertAuth: true,
+}
+
+func TestDiscoverInitialCluster(t *testing.T) {
+	clus := integration.NewClusterV3(t, &integration.ClusterConfig{
+		Size:               3,
+		PeerTLS:            &testTLSInfo,
+		ClientTLS:          &testTLSInfo,
+		SkipCreatingClient: true,
+	})
+	defer clus.Terminate(t)
+
+	target, err := url.Parse(clus.Members[0].PeerURLs.String())
+	if err != nil {
+		t.Fatal(err)
+	}
+	targertHost, targetPort, err := net.SplitHostPort(target.Host)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	wantInitialCluster := []string{}
+	for _, member := range clus.Members {
+		// generate initial cluster <name>=<peerURL>
+		wantInitialCluster = append(wantInitialCluster, fmt.Sprintf("%s=%s", member.Name, member.PeerURLs.String()))
+	}
+
+	tests := map[string]struct {
+		targetName         string
+		endpoints          []string
+		dataDir            string
+		wantInitialCluster []string
+		wantErr            bool
+	}{
+		"member found no data-dir": {
+			targetName:         clus.Members[0].Name,
+			endpoints:          []string{clus.Members[0].GRPCAddr(), clus.Members[1].GRPCAddr(), clus.Members[2].GRPCAddr()},
+			dataDir:            "",
+			wantInitialCluster: wantInitialCluster,
+		},
+		"member found with data-dir": {
+			targetName:         clus.Members[0].Name,
+			endpoints:          []string{clus.Members[0].GRPCAddr(), clus.Members[1].GRPCAddr(), clus.Members[2].GRPCAddr()},
+			dataDir:            clus.Members[0].DataDir,
+			wantInitialCluster: wantInitialCluster,
+		},
+		"timeout no dataDir": { // expect retry cluster member has never started
+			targetName:         clus.Members[0].Name,
+			endpoints:          []string{"http://127.0.0.1:1234"},
+			dataDir:            "",
+			wantInitialCluster: []string{""},
+			wantErr:            true,
+		},
+		"timeout with data-dir": { // if the dataDir exists the member has started in the past, try to start.
+			targetName:         clus.Members[0].Name,
+			endpoints:          []string{"http://127.0.0.1:1234"},
+			dataDir:            clus.Members[0].DataDir,
+			wantInitialCluster: []string{""},
+			wantErr:            false,
+		},
```

Comment:
Should we add a test case for `member not found with data-dir`? (on and off scenarios L175 and L145)
https://github.com/openshift/etcd/pull/73/files#diff-3d1414e0d47047e4fdcae958fb7654d082c1eb1b71cf4eb38aeb3190db678208R178

Reply:
yeah I think that would be another bug I mean we restart the process but we should allow etcd time to try and start.